pymeasure/instruments/thyracont/smartline_v2.py

```diff
@@ -0,0 +1,505 @@
+
+    Sources = Sources
+
+    errors = {'NO_DEF': "Invalid command for this device.",
+              '_LOGIC': "Access Code is invalid or illogical command.",
+              '_RANGE': "Value sent is out of range.",
+              'ERROR1': "Sensor defect or stacked out.",
+              'SYNTAX': "Wrong syntax or mode in data is invalid for this device.",
+              'LENGTH': "Length of data is out of expected range.",
+              '_CD_RE': "Calibration Data Read Error.",
+              '_EP_RE': "EEPROM Read Error.",
+              '_UNSUP': "Unsupported Data for that command.",
+              '_SEDIS': "Sensor element disabled."}
+
+    def __init__(self, adapter, name="Thyracont SmartlineV2 Transmitter", baud_rate=115200,
+                 address=1, timeout=250,
+                 **kwargs):
+        super().__init__(adapter, name=name, includeSCPI=False,
+                         write_termination="\r",
+                         read_termination="\r",
+                         timeout=timeout,
+                         asrl={'baud_rate': baud_rate,
+                               'parity': Parity.none,
+                               'stop_bits': StopBits.one,
+                               },
+                         **kwargs
+                         )
+        self.address = address  # 1-16
+
+    def write(self, command):
+        """Write a command to the device."""
+        message = f"{self.address:03}{command}"
+        super().write(f"{message}{calculateChecksum(message)}")
+
+    def write_composition(self, accessCode, command, data=""):
+        """Write a command with an accessCode and optional data to the device.
+
+        :param accessCode: How to access the device.
+        :param command: Two char command string to send to the device.
+        :param data: Data for the command.
+        """
+        self.write(f"{accessCode}{command}{compose_data(data)}")
+
+    def ask(self, command_message, query_delay=0):
+        """Ask for some value and check that the response matches the original command.
+
+        :param str command_message: Access code, command, length, and content.
+            The command sent is compared to the response command.
+        """
+        self.write(command_message)
+        self.wait_for(query_delay)
+        return self.read(command_message[1:3])
+
+    def ask_manually(self, accessCode, command, data="", query_delay=0):
+        """
+        Send a message to the transmitter and return its answer.
+
+        :param accessCode: How to access the device.
+        :param command: Command to send to the device.
+        :param data: Data for the command.
+        :param int query_delay: Time to wait between writing and reading.
+        :return str: Response from the device after error checking.
+        """
+        self.write(f"{accessCode}{command}{compose_data(data)}")
+        self.wait_for(query_delay)
+        return self.read(command)
+
+    def read(self, command=None):
```

Comment:
I think (not sure though), you can rewrite the `read` to call `super` reading up to termination character and then filter out unwanted values or use ignore errors in `decode`.

Reply:
Thanks for the hint at `errors="ignore"`.
It does not solve the zero byte (b"\x00") problem, but that can be fixed with a simple replace.

In order to verify correct transition of the code, I installed a test.